code/drasil-lang/Language/Drasil/Space.hs

```diff
@@ -25,9 +25,20 @@ data Space =
   | Actor String
   | DiscreteD [Double]
   | DiscreteS [String] --ex. let Meal = {"breakfast", "lunch", "dinner"}
+  | Mapping [Space] Space -- TODO: this one shouldn't be exposed directly, only through a smart constructor
   | Void
   deriving (Eq, Show)
 
+mkPrimitiveMapping :: [Space] -> Space -> Space
+mkPrimitiveMapping _   (Mapping _ _) = error "Target must be basic"
```

Comment:
So rather than this, it would make sense to create a `Primitive` type that is just a wrapper for `Space`, but then use it in the above, i.e.
```haskell
   | Mapping [Primitive] Primitive
```
so that the invariant is also "in the type".

Plus, `mkFunction` is a better name...

Reply:
Sounds good. What do you think of creating a PrimitiveSpace ADT for the primitives? Something along the lines of:
```
data PrimitiveSpace = Integer | Rational | Real | Natural | Boolean | ...
data Space = Primitive PrimitiveSpace | Function (NonEmpty PrimitiveSpace) PrimitiveSpace
```
I think this would also resolve the boolean blindness issue below. However, this would still be insufficient for the function spaces we need for #2395.

Reply:
I renamed `Mapping` to `Function` as well.

Reply:
`PrimitiveSpace` is a good idea.